File: Problems/Algorithms/290. Word Pattern/WordPattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        Set<String> wordSet = new HashSet();
        Set<Character> patternSet = new HashSet();
        
        String[] words = s.split(" ");
        for (String word : words) wordSet.add(word);
        for (char ch : pattern.toCharArray()) patternSet.add(ch);
        
        if (words.length != pattern.length() || wordSet.size() != patternSet.size()) return false;
        
        Map<Character, String> map = new HashMap<>();
        for (int i = 0; i < words.length; i++) {
            char key = pattern.charAt(i);
            if (map.containsKey(key)) {
                if (!map.get(key).equals(words[i])) return false;
            } else {
                map.put(key, words[i]);
            }
        }
        
        return true;
    }
}

File: Problems/Algorithms/290. Word Pattern/word_pattern.rs
use std::collections::{HashMap, HashSet};

impl Solution {
    pub fn word_pattern(pattern: String, s: String) -> bool {
        if pattern.len() != s.matches(' ').count() + 1 {
            return false;
        }

        let mut map = HashMap::new();
        let mut words = HashSet::new();
        for (word, character) in s.split_ascii_whitespace().zip(pattern.chars()) {
            if let Some(existing_word) = map.insert(character, word) {
                if existing_word != word {
                    return false;
                }
            } else if !words.insert(word) {
                return false;
            }
        }

        true
    }
}
